File: 2019/day_12.R
# https://adventofcode.com/2019/day/12


# At first I struggled with part 2 but then I found this comment on Reddit:
# https://www.reddit.com/r/adventofcode/comments/e9j0ve/2019_day_12_solutions/faja0lj
# The key insights are:
# 1) The axes (x,y,z) are totally independent. So it suffices to find the period
# for each axis separately. Then the answer is the lcm of these.
# 2) Each axis will repeat "relatively quickly" (fast enough to brute force)
# 3) Since each state has a unique parent, the first repeat must be a repeat of
# state 0.


positions <- readLines("2019/day_12_input.txt")
positions <- regmatches(positions, gregexpr("-?\\d+", positions))
positions <- do.call(rbind, positions)
class(positions) <- "numeric"
moons <- cbind(positions, matrix(0, ncol = 3, nrow = 4))
colnames(moons) <- c("x", "y", "z", "vx", "vy", "vz")
moons

x_original <- moons[, c("x", "vx")]
y_original <- moons[, c("y", "vy")]
z_original <- moons[, c("z", "vz")]
x_period <- NA
y_period <- NA
z_period <- NA

moon_pairs <- t(combn(1:4, 2))
n_pairs <- nrow(moon_pairs)

pos <- 1:3
vel <- 4:6
n_energy <- 1000
i <- 0
while (TRUE) {
    i <- i + 1

    # apply gravity
    for (p in seq_len(n_pairs)) {
        a <- moon_pairs[p, 1]
        b <- moon_pairs[p, 2]
        a_greater_b <- moons[a, pos] > moons[b, pos]
        b_greater_a <- moons[b, pos] > moons[a, pos]
        moons[a, vel][a_greater_b] <- moons[a, vel][a_greater_b] - 1
        moons[b, vel][a_greater_b] <- moons[b, vel][a_greater_b] + 1
        moons[a, vel][b_greater_a] <- moons[a, vel][b_greater_a] + 1
        moons[b, vel][b_greater_a] <- moons[b, vel][b_greater_a] - 1
    }

    # apply velocity
    moons[, pos] <- moons[, pos] + moons[, vel]

    if (i == n_energy) {
        energy <- sum(apply(abs(moons), 1, function(x) sum(x[pos]) * sum(x[vel])))
    }

    if (is.na(x_period) && identical(moons[, c(1, 4)], x_original)) {
        x_period <- i
    }
    if (is.na(y_period) && identical(moons[, c(2, 5)], y_original)) {
        y_period <- i
    }
    if (is.na(z_period) && identical(moons[, c(3, 6)], z_original)) {
        z_period <- i
    }

    if (i %% 10000 == 0) cat("step", i, "\n")
    if (i >= n_energy && !any(is.na(c(x_period, y_period, z_period)))) break
}


gcd <- function(a, b) {
    while (b > 0) {
        temp = b
        b = a %% b
        a = temp
    }
    a
}

lcm <- function(a, b) {
    a * b / gcd(a, b)
}


# part 1:
print(energy)  # 8742

# part 2:
match_step <- lcm(lcm(x_period, y_period), z_period)
print(format(match_step, scientific = FALSE))  # 325433763467176

File: src/year2019/day12.py
# https://adventofcode.com/2019/day/12

# At first, I struggled with part 2, but then I found this comment on Reddit:
# https://www.reddit.com/r/adventofcode/comments/e9j0ve/2019_day_12_solutions/faja0lj
# The key insights are:
# 1) The axes (x,y,z) are totally independent. So it suffices to find the period
# for each axis separately. Then the answer is the least common multiple of these.
# 2) Each axis will repeat "relatively quickly" (fast enough to brute force)
# 3) Since each state has a unique parent, the first repeat must be a repeat of state 0.

import itertools
import math
import re

from src.util.types import Data, Solution


COMBINATIONS = list(itertools.combinations(range(4), 2))


def prepare_data(data: str) -> list[list[int]]:
    pattern = re.compile(r"-?\d+")
    return [[int(n) for n in re.findall(pattern, line)] for line in data.splitlines()]


def simulate(data: list[list[int]], n_steps: int) -> tuple[int, int]:
    x_positions = [d[0] for d in data]
    y_positions = [d[1] for d in data]
    z_positions = [d[2] for d in data]
    x_velocities = [0] * 4
    y_velocities = [0] * 4
    z_velocities = [0] * 4

    x_positions_original = x_positions.copy()
    y_positions_original = y_positions.copy()
    z_positions_original = z_positions.copy()
    x_velocities_original = x_velocities.copy()
    y_velocities_original = y_velocities.copy()
    z_velocities_original = z_velocities.copy()

    total_energy = 0
    x_period = 0
    y_period = 0
    z_period = 0

    i = 0
    while i < n_steps or x_period == 0 or y_period == 0 or z_period == 0:
        i += 1

        if x_period == 0 or i <= n_steps:
            apply_gravity(x_positions, x_velocities)
            if x_positions == x_positions_original and x_velocities == x_velocities_original:
                x_period = i
        if y_period == 0 or i <= n_steps:
            apply_gravity(y_positions, y_velocities)
            if y_positions == y_positions_original and y_velocities == y_velocities_original:
                y_period = i
        if z_period == 0 or i <= n_steps:
            apply_gravity(z_positions, z_velocities)
            if z_positions == z_positions_original and z_velocities == z_velocities_original:
                z_period = i

        if i == n_steps:
            for j in range(4):
                potential = sum(abs(n) for n in (x_positions[j], y_positions[j], z_positions[j]))
                kinetic = sum(abs(n) for n in (x_velocities[j], y_velocities[j], z_velocities[j]))
                total_energy += potential * kinetic

    return total_energy, math.lcm(x_period, y_period, z_period)


def apply_gravity(positions: list[int], velocities: list[int]) -> None:
    for i, j in COMBINATIONS:
        if positions[i] > positions[j]:
            velocities[i] -= 1
            velocities[j] += 1
        elif positions[i] < positions[j]:
            velocities[i] += 1
            velocities[j] -= 1
    for i in range(4):
        positions[i] += velocities[i]


def solve(data: Data) -> Solution:
    solution = Solution()
    sample_data_0 = prepare_data(data.samples[0])
    sample_data_1 = prepare_data(data.samples[1])
    sample_solutions_0 = simulate(sample_data_0, 10)
    sample_solutions_1 = simulate(sample_data_1, 100)
    solution.samples_part_1.append(sample_solutions_0[0])
    solution.samples_part_1.append(sample_solutions_1[0])
    solution.samples_part_2.append(sample_solutions_0[1])
    solution.samples_part_2.append(sample_solutions_1[1])

    challenge_data = prepare_data(data.input)
    energy, period = simulate(challenge_data, 1000)
    solution.part_1 = energy
    solution.part_2 = period
    return solution
